File: UnitySample/Assets/Samples/SteamLogin/SteamLoginTest.cs
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Networking;
using UnityEngine.UI;

public class SteamLoginTest : MonoBehaviour
{
    public string loginEndpointUrl;
    public Text logOutput;

    // Start is called before the first frame update
    void Start()
    {
        // Set the login endpoint
        Debug.Log("Setting login endpoint");
        AWSGameSDKClient.Instance.Init(loginEndpointUrl, this.OnLoginOrRefreshError);

        // Get a new guest user first to test account linking
        Debug.Log("Requesting new guest identity first");
        this.logOutput.text += "Requesting new identity\n";
        AWSGameSDKClient.Instance.LoginAsNewGuestUser(this.OnGuestLoginResponse);
    }

    // Triggered by the SDK if there's any login error or error refreshing access token
    void OnLoginOrRefreshError(string error)
    {
        Debug.LogError("Login or refresh error: " + error);
        this.logOutput.text += "Login or refresh error: " + error + "\n";

        // NOTE: You would here trigger a new log in or other remediation
    }

    // Update is called once per frame
    void Update()
    {
        
    }

    void OnGuestLoginResponse(LoginRequestData userInfo)
    {
        Debug.Log("New guest Login response: UserID: " + userInfo.user_id + "GuestSecret: " + userInfo.guest_secret + "Token: " + userInfo.auth_token);
        this.logOutput.text += "Login response: \nUserID: " + userInfo.user_id + " \nGuestSecret: " + userInfo.guest_secret + " \n";
        
        // Now let's test linking Steam ID to the guest user
        // NOTE: You're expected to input a valid Steam token here, see the documentation for Unity Steam integration here: https://steamworks.github.io/
        //       Also see the Steam documentation for using GetAuthTicketForWebApi (new) and GetAuthSessionTicket (old): https://partner.steamgames.com/doc/features/auth
        AWSGameSDKClient.Instance.LinkSteamIdToCurrentUser("YourTokenHere", this.OnLinkSteamIdResponse);
    }

    void OnLinkSteamIdResponse(LoginRequestData userInfo)
    {
        Debug.Log("Steam ID linking response: UserID: " + userInfo.user_id + "SteamId: " + userInfo.steam_id + "Token: " + userInfo.auth_token);
        this.logOutput.text += "Steam ID linking response: \nUserID: " + userInfo.user_id + "\nSteamId: " + userInfo.steam_id + " \n";

        // Now let's test logging in with this existing Steam ID
        // NOTE: As above, you're expected to have a valid Steam token here
        AWSGameSDKClient.Instance.LoginWithSteamToken("YourTokenHere", this.OnLoginWithSteamResponse);
    }

    void OnLoginWithSteamResponse(LoginRequestData userInfo)
    {
        Debug.Log("Steam ID login response: UserID: " + userInfo.user_id + "SteamId: " + userInfo.steam_id + "Token: " + userInfo.auth_token);
        this.logOutput.text += "Steam ID login response: \nUserID: " + userInfo.user_id + "\nSteamId: " + userInfo.steam_id + " \nToken: " + userInfo.auth_token + "\n";
    }
}

File: UnrealSample/Source/UnrealSample/Samples/SteamLoginTest/SteamLoginTest.cpp
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

#include "SteamLoginTest.h"
#include "../../AWSGameSDK/AWSGameSDK.h"
#include "Engine/World.h"
#include "Kismet/GameplayStatics.h"

// Sets default values for this component's properties
USteamLoginTest::USteamLoginTest()
{
	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
	// off to improve performance if you don't need them.
	PrimaryComponentTick.bCanEverTick = true;

	// ...
}


// Called when the game starts
void USteamLoginTest::BeginPlay()
{
	Super::BeginPlay();

	// Get the subsystems
    UGameInstance* GameInstance = Cast<UGameInstance>(UGameplayStatics::GetGameInstance(GetWorld()));
    UAWSGameSDK* AWSGameSDK =  GameInstance->GetSubsystem<UAWSGameSDK>();

    // Init with the login endpoint defined in the Editor and a callback to handle errors for logging in and refresh
	AWSGameSDK->Init(this->m_loginEndpoint);
	AWSGameSDK->OnLoginFailure.AddUObject(this, &USteamLoginTest::OnLoginOrRefreshErrorCallback);

	// Define the OnLoginResult callback
	UAWSGameSDK::FLoginComplete loginCallback;
	loginCallback.BindUObject(this, &USteamLoginTest::OnGuestLoginResultCallback);

	// Login as a new guest user first
    AWSGameSDK->LoginAsNewGuestUser(loginCallback);
	
}

// Called when there is an error with login or token refresh. You will need to handle logging in again here
void USteamLoginTest::OnLoginOrRefreshErrorCallback(const FString& errorMessage){
    UE_LOG(LogTemp, Display, TEXT("Received login error: %s \n"), *errorMessage);
    if(GEngine)
        GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Red, FString::Printf(TEXT("Received login error: \n %s \n"), *errorMessage), false, FVector2D(1.5f,1.5f));

    // NOTE:  You will need to handle logging in again here
}

// Called when guest login is done
void USteamLoginTest::OnGuestLoginResultCallback(const UserInfo& userInfo){
    UE_LOG(LogTemp, Display, TEXT("Received guest login response: %s \n"), *userInfo.ToString());
    if(GEngine)
            GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Black, FString::Printf(TEXT("Received login response: \n %s \n"), *userInfo.ToString()), false, FVector2D(1.5f,1.5f));

    // Test linking steam id to the existing guest account
	UAWSGameSDK::FLoginComplete onLinkSteamIdCallback;
	onLinkSteamIdCallback.BindUObject(this, &USteamLoginTest::OnLinkSteamIdResultCallback);
	UGameInstance* GameInstance = Cast<UGameInstance>(UGameplayStatics::GetGameInstance(GetWorld()));
    UAWSGameSDK* AWSGameSDK = GameInstance->GetSubsystem<UAWSGameSDK>();

    // NOTE: You need a Steam token here retrieved with either GetAuthTicketForWebApi (new) or GetAuthSessionTicket (old): https://partner.steamgames.com/doc/features/auth
    //       See Unreal Steam Online subsystem documentation for more details on integration: https://docs.unrealengine.com/5.2/en-US/online-subsystem-steam-interface-in-unreal-engine/
	AWSGameSDK->LinkSteamIdToCurrentUser("tokenHere", onLinkSteamIdCallback);
}

void USteamLoginTest::OnLinkSteamIdResultCallback(const UserInfo& userInfo){

    UE_LOG(LogTemp, Display, TEXT("Received Steam ID linking response: %s \n"), *userInfo.ToString());
    if(GEngine)
            GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Black, FString::Printf(TEXT("Received Steam ID linking response: \n %s \n"), *userInfo.ToString()), false, FVector2D(1.5f,1.5f));

    // Test logging in with existing steam_id
	UAWSGameSDK::FLoginComplete onSteamIdLoginCallback;
	onSteamIdLoginCallback.BindUObject(this, &USteamLoginTest::OnLoginWithSteam);
	UGameInstance* GameInstance = Cast<UGameInstance>(UGameplayStatics::GetGameInstance(GetWorld()));
    UAWSGameSDK* AWSGameSDK = GameInstance->GetSubsystem<UAWSGameSDK>();

    // NOTE: As above, you're expected to have a valid Steam token here
    AWSGameSDK->LoginWithSteamToken("tokenHere", onSteamIdLoginCallback);
}

void USteamLoginTest::OnLoginWithSteam(const UserInfo& userInfo){

    UE_LOG(LogTemp, Display, TEXT("Received Steam login response: %s \n"), *userInfo.ToString());
    if(GEngine)
        GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Black, FString::Printf(TEXT("Received Steam login response: \n %s \n"), *userInfo.ToString()), false, FVector2D(1.5f,1.5f));
}

// Called every frame
void USteamLoginTest::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
{
	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);

	// ...
}

